### PR TITLE
Model Loading and Conversion

### PR DESCRIPTION
**Pull request by alberthli**
This PR adds basic utils for loading `mjx` models from URDFs/XMLs. It also demonstrates how to convert URDFs to `mujoco`-compliant XMLs using only the python API.

### CODE REVIEW

File: ambersim/utils.py
Comment (kli58) [inline ambersim/utils.py:1]: We need a function that adds a freejoint to existing model (if loading from urdf) and a flag that allows the user to determine whether this is a floating-based robot or not. 
See the reference here: https://github.com/unitreerobotics/unitree_mujoco
We could either modify the xml string directly or use addBody to add the freejoint

File: tests/test_model_io.py
Comment (kli58) [inline tests/test_model_io.py:1]: see comment on ambersim/utils.py. Please add test case for floating-base model

File: ambersim/utils.py
Comment (alberthli) [reply ambersim/utils.py:1]: addressed by modifying the xml string using `dm_control` from deepmind. See `_modify_robot_float_base`.

File: tests/test_model_io.py
Comment (alberthli) [reply tests/test_model_io.py:1]: added two test cases: (1) if the model already has any joint (e.g., free, revolute, etc.) attached to its first non-world body, then don't add a freejoint; (2) if it doesn't have any joints there, then edit the xml string and load a freejoint.

see: `test_force_float`

Comment (kli58) [reply tests/test_model_io.py:1]: Please add test case to make sure the loaded decomposed convex meshes have the same face,vert,edge as defined in geom_convex_face, geom_convex_vert, geom_convex_edge for collision checking (see reference here: https://github.com/google-deepmind/mujoco/blob/57e6940f579484adf34eebedc51279a818909f34/mjx/mujoco/mjx/_src/mesh.py)

File: ambersim/utils/io_utils.py
Comment (kli58) [inline ambersim/utils/io_utils.py:1]: Please add test case to make sure the loaded decomposed convex meshes have the same face,vert,edge as defined in geom_convex_face, geom_convex_vert, geom_convex_edge for collision checking (see reference here: https://github.com/google-deepmind/mujoco/blob/57e6940f579484adf34eebedc51279a818909f34/mjx/mujoco/mjx/_src/mesh.py)

File: ambersim/utils/io_utils.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+def mj_to_mjx_model(mj_model: mj.MjModel) -> Tuple[mjx.Model, mjx.Data]:
+    """Converts a mujoco model to an mjx (model, data) pair."""
+    try:
+        mjx_model = mjx.device_put(mj_model)
+        mjx_data = mjx.make_data(mj_model)
+        return mjx_model, mjx_data
+    except NotImplementedError as e:
+        print()
+        print("There are some URDF convex primitives that aren't compatible with mjx's convex collision checking.")
```
Comment (alberthli) [inline ambersim/utils/io_utils.py]: TODO(ahl): As per offline conversation with @pculbertson, this extended print statement should be expanded to include other possible causes for throwing the error, including friction cone stuff. 

Reference this feature parity table: https://mujoco.readthedocs.io/en/latest/mjx.html#feature-parity

File: ambersim/utils/io_utils.py
Comment (alberthli) [reply ambersim/utils/io_utils.py:1]: nit: in the future, add these comments closer to the actual line where the change is going to be made so it's easier to track (i.e., this comment should have been added in the file with tests)

File: tests/test_model_io.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        tm = trimesh.Trimesh(vertices=dverts, faces=dfaces)
+        tm_convex = trimesh.convex.convex_hull(tm)
+        tverts = tm_convex.vertices
+        tfaces = tm_convex.faces
+
+        # checking that signed distance queries are functionally the same
+        coords = np.random.randn(100, 3)
+        signed_dist_d = igl.signed_distance(coords, dverts, dfaces)[0]
+        signed_dist_t = igl.signed_distance(coords, tverts, tfaces)[0]
+        assert np.allclose(signed_dist_d, signed_dist_t)
```
Comment (alberthli) [inline tests/test_model_io.py:127]: @kli58 is this enough for the test case? 

In [these lines](https://github.com/google-deepmind/mujoco/blob/57e6940f579484adf34eebedc51279a818909f34/mjx/mujoco/mjx/_src/mesh.py#L195C23-L207), the `tm_convex` object gets fed into the `mjx` ecosystem with their utils, but I'm not actually testing against the `mjx` outputs, since we can't prevent it from preprocessing our input geoms anyway.

File: ambersim/utils/io_utils.py (hunk 1 above)
Comment (alberthli) [reply ambersim/utils/io_utils.py]: Done.

File: tests/test_model_io.py (hunk 2 above)
Comment (alberthli) [reply tests/test_model_io.py:127]: reference to `igl`'s `signed_distance` function: https://libigl.github.io/libigl-python-bindings/igl_docs/#signed_distance

File: ambersim/utils/io_utils.py
Comment (kli58) [inline ambersim/utils/io_utils.py:1]: For ```_modify_robot_float_base```, looks like you are adding default inertial values, these should come from urdf if specified there right?

File: ambersim/utils/io_utils.py [hunk 3]
```diff
@@ -47,57 +48,80 @@ def _modify_robot_float_base(filepath: Union[str, Path]) -> mj.MjModel:
...
+        mj_model = mj.MjModel.from_xml_path(filepath)
+    return mj_model
+
 
+def mj_to_mjx_model(mj_model: mj.MjModel) -> Tuple[mjx.Model, mjx.Data]:
+    """Converts a mujoco model to an mjx (model, data) pair."""
     try:
-        mjx_model = mjx.device_put(_model)
-        mjx_data = mjx.make_data(_model)
+        mjx_model = mjx.device_put(mj_model)
```
Comment (vincekurtz) [inline ambersim/utils/io_utils.py:91]: When I run this I get warnings about MJX not supporting the Newton solver. It looks from the [website](https://mujoco.readthedocs.io/en/latest/mjx.html) like MJX does support Newton now, but I wasn't able to get that version from pip. Do we want to force MJX to use the CG solver, e.g., `mj_model.opt.solver = mj.mjtSolver.mjSOL_CG`?

Comment (kli58) [reply ambersim/utils/io_utils.py:91]: It should by default force it to CG solver right now on MJX side. But they also have already implemented on Newton solver it's just not on their stable release yet

Comment (alberthli) [reply ambersim/utils/io_utils.py:91]: I opened an issue last night about allowing `mujoco` install from source instead of vs. `pip` because the nightly release has a lot of new upgrades over `3.0.0` including support for the Newton solver. I basically never want to force the CG solver once this becomes the official `pypi` release, so maybe this can get resolved in a separate PR?

In the meantime, re-installing `mujoco` from github on your local env directly should kill the warning.

Comment (vincekurtz) [reply ambersim/utils/io_utils.py:91]: Right, just wondering if we should make that conversion explicit so we don't get warnings when we run `pytest`. I want those pretty green checkmarks!

Comment (alberthli) [reply ambersim/utils/io_utils.py:91]: I can expose the option for the solver in the API and add a TODO to come back and make Newton the default once the other PR is done + add this to that issue

File: ambersim/utils/io_utils.py
Comment (alberthli) [reply ambersim/utils/io_utils.py:1]: I opened an issue about `mujoco` deleting base links of URDF-loaded robots: https://github.com/google-deepmind/mujoco/issues/1176

Hopefully this gives some insight into our offline discussion and can help us resolve the parsing issue which makes the above hack necessary.

In the meantime, I'll try thinking of a workaround by parsing the URDF directly and extracting the base link...

File: ambersim/utils/io_utils.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    # only add free joint if the first body after worldbody has no freejoints
+    assert robot.worldbody is not None
+    if robot.worldbody.body[0].freejoint is None:
+        # creates a new robot object and attaches the current robot to it freely
+        arena = mjcf.RootElement(model=robot.model)
+        attachment_frame = arena.attach(robot)
+        attachment_frame.add("freejoint", name="freejoint")
+        robot = arena
+
+        # ensuring that the base link has inertial attributes
```
Comment (alberthli) [inline ambersim/utils/io_utils.py:62]: @kli58 (I moved this comment here for readability)

Minor comment on ```mj_to_mjx_model```: this read as only returning ```mjx_mode```l instead of (```mjx_model,mjx_data```) pair

File: ambersim/utils/io_utils.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            mj_model = _modify_robot_float_base(output_name)
+            Path.unlink(output_name)
+        else:
+            mj_model = _modify_robot_float_base(filepath)
+    else:
+        mj_model = mj.MjModel.from_xml_path(filepath)
+    return mj_model
+
+
+def mj_to_mjx_model(mj_model: mj.MjModel) -> Tuple[mjx.Model, mjx.Data]:
```
Comment (alberthli) [inline ambersim/utils/io_utils.py]: @kli58 (I moved this comment here for readability)

Minor comment on ```mj_to_mjx_model```: this read as only returning ```mjx_mode```l instead of (```mjx_model,mjx_data```) pair

Comment (alberthli) [reply ambersim/utils/io_utils.py]: Updated function name.

File: ambersim/utils/io_utils.py (hunk 3 above)
Comment (alberthli) [reply ambersim/utils/io_utils.py:91]: Exposed:
```
def load_mj_model_from_file(
    filepath: Union[str, Path],
    force_float: bool = False,
    solver: Union[str, mj.mjtSolver] = mj.mjtSolver.mjSOL_CG,
    iterations: Optional[int] = None,
    ls_iterations: Optional[int] = None,
) -> mj.MjModel:
    """Loads a mujoco model from a filepath.
    Args:
        filepath: A path to a URDF or MJCF file. This can be global, local, or with respect to the repository root.
        force_float: Whether to forcibly float the base of the robot.
        solver: The type of solver to use. For now, mjx only supports CG.
        iterations: The number of solver iterations to use. If unspecified, mujoco uses 100.
        ls_iterations: The number of line search iterations to use. If unspecified, mujoco uses 50.
    Returns:
        mj_model: A mujoco model.
    """
    # TODO(ahl): once we allow installing mujoco from source, update this to allow the Newton solver + update default
    if isinstance(solver, str):
        if solver.lower() == "cg":
            solver = mj.mjtSolver.SOL_CG
        else:
            raise ValueError("Solver must be one of: ['cg']!")
    elif isinstance(solver, mj.mjtSolver):
        assert solver in [mj.mjtSolver.mjSOL_CG]
```

File: ambersim/utils/io_utils.py
Comment (alberthli) [reply ambersim/utils/io_utils.py:1]: As per discussion, the `fusedstatic="false"` model compiler flag gives us the desired behavior for correctly loading URDFs. This has been added in the `mujoco` blocks of all example URDFs.

We also discussed adding functionality for exposing these settings programmatically so that the end user doesn't have to edit their URDFs. I'm of the opinion that this feature should go in a separate PR, since this one is getting pretty big.

File: ambersim/utils/io_utils.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    """Modifies a robot to have a floating base if it doesn't already."""
+    # loading current robot
+    assert str(filepath).split(".")[-1] == "xml"
+    robot = mjcf.from_file(filepath, model_dir="/".join(filepath.split("/")[:-1]))
+
+    # only add free joint if the first body after worldbody has no freejoints
+    assert robot.worldbody is not None
+    if len(robot.worldbody.body[0].joint) == 0:
+        robot.worldbody.body[0].add("freejoint", name="freejoint")
+        assert robot.worldbody.body[0].inertial is not None  # checking for non-physical parsing errors
```
Comment (alberthli) [inline ambersim/utils/io_utils.py:26]: @kli58 this is where the inertial properties were previously checked. For now, we rely on the user correctly specifying `fusestatic` in the URDF - if so, this assertion will pass.

**Review by kli58 — APPROVED** [review]
